components/raftstore-v2/src/operation/command/write/mod.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use engine_traits::{KvEngine, RaftEngine};
+use kvproto::raft_cmdpb::RaftCmdRequest;
+use raftstore::{
+    store::{
+        cmd_resp,
+        fsm::{apply, Proposal},
+        msg::ErrorCallback,
+        WriteCallback,
+    },
+    Result,
+};
+use tikv_util::Either;
+
+use crate::{batch::StoreContext, raft::Peer, router::CmdResChannel};
+
+mod raw_write;
+
+pub use raw_write::{RawWriteDecoder, RawWriteEncoder};
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    #[inline]
+    pub fn on_write_command<T>(
+        &mut self,
+        ctx: &mut StoreContext<EK, ER, T>,
+        mut req: RaftCmdRequest,
+        ch: CmdResChannel,
+    ) {
+        if !self.serving() {
+            apply::notify_req_region_removed(self.region_id(), ch);
+            return;
+        }
+        if let Some(encoder) = self.raw_write_encoder_mut() {
+            match encoder.amend(req) {
```

Comment:
This relies on a strong assumption that amended requests is always valid. It requires `validate_command` only checks header and cmd type, and the state of this fsm is unchanged.

Reply:
After being proposed, there can be only three types of failure: epoch not matched, stale command due to leadership change, key not in range. Because headers of all batched commands are the same, there will be no extra commands failure due to first reason. For the second reason, it can happen no matter batch or not. For the third reason, it's possible an invalid command can make correct command fail. But generally, a correctly implemented client should always check epoch first and then check range, so it's impossible to trigger such error.

So the assumption here is all the clients are implemented correctly.